File: rust/src/lexer.rs
use core::panic;

use crate::token::{token_init, TokenType};

#[derive(Debug)]
pub struct LexerImpl {
    pub tokens: Vec<crate::token::Token>,
    start: usize,
    curr: usize,
    line: u32,
    source: String,
}

pub fn lexer_init(source: String) -> LexerImpl {
    LexerImpl {
        tokens: vec![],
        start: 0,
        curr: 0,
        line: 1,
        source,
    }
}

pub trait Lexer {
    fn tokenize(&mut self);
    fn advance(&mut self) -> char;
    fn add_token(&mut self, token_type: TokenType);
    fn peek(&self) -> char;
    fn lookahead(&self, n: usize) -> char;
    fn match_char(&mut self, expected: char) -> bool;
    fn handle_number(&mut self);
    fn handle_string(&mut self, quote: char);
    fn handle_identifier(&mut self);
}

impl Lexer for LexerImpl {
    fn advance(&mut self) -> char {
        if self.curr >= self.source.len() {
            return '\0';
        }
        let ch = self.source.chars().nth(self.curr);
        self.curr += 1;
        ch.unwrap()
    }

    fn add_token(&mut self, token_type: TokenType) {
        self.tokens.push(token_init(
            token_type,
            self.source[self.start..self.curr].to_owned(),
            self.line,
        ));
    }

    fn peek(&self) -> char {
        if self.curr >= self.source.len() {
            return '\0';
        }
        self.source.chars().nth(self.curr).unwrap()
    }

    fn lookahead(&self, n: usize) -> char {
        if self.curr + n >= self.source.len() {
            return '\0';
        }
        self.source.chars().nth(self.curr + n).unwrap()
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.curr >= self.source.len() || self.peek() != expected {
            return false;
        }
        self.curr += 1;
        true
    }

    fn handle_number(&mut self) {
        while self.peek().is_ascii_digit() {
            self.advance();
        }
        if self.peek() == '.' && self.lookahead(1).is_ascii_digit() {
            self.advance();
            while self.peek().is_ascii_digit() {
                self.advance();
            }
            self.add_token(TokenType::TokFloat);
        } else {
            self.add_token(TokenType::TokInteger);
        }
    }

    fn handle_string(&mut self, quote: char) {
        while self.peek() != quote && self.curr < self.source.len() {
            self.advance();
        }
        if self.curr >= self.source.len() {
            panic!("Unfinished string !{}", self.line);
        }
        self.advance();
        self.add_token(TokenType::TokString);
    }

    fn handle_identifier(&mut self) {
        while self.peek().is_alphanumeric() || self.peek() == '_' {
            self.advance();
        }
        self.add_token(crate::token::KEYWORDS.with(|keyword| {
            *keyword
                .get(&self.source[self.start..self.curr])
                .unwrap_or(&TokenType::TokIdentifier)
        }));
    }

    fn tokenize(&mut self) {
        while self.curr < self.source.len() {
            self.start = self.curr;
            let ch = self.advance();
            match ch {
                '\n' => self.line += 1,
                '#' => {
                    while self.peek() != '\n' && self.curr < self.source.len() {
                        self.advance();
                    }
                }
                '(' => self.add_token(TokenType::TokLparen),
                ')' => self.add_token(TokenType::TokRparen),
                '{' => self.add_token(TokenType::TokLcurly),
                '}' => self.add_token(TokenType::TokRcurly),
                '[' => self.add_token(TokenType::TokLsquar),
                ']' => self.add_token(TokenType::TokRsquar),
                ',' => self.add_token(TokenType::TokComma),
                '.' => self.add_token(TokenType::TokDot),
                '+' => self.add_token(TokenType::TokPlus),
                '-' => {
                    if self.match_char('-') {
                        while self.peek() != '\n' && self.curr < self.source.len() {
                            self.advance();
                        }
                    } else {
                        self.add_token(TokenType::TokMinus);
                    }
                }
                '*' => self.add_token(TokenType::TokStar),
                '/' => self.add_token(TokenType::TokSlash),

                '^' => self.add_token(TokenType::TokCaret),
                '%' => self.add_token(TokenType::TokMod),
                ':' => {
                    if self.match_char('=') {
                        self.add_token(TokenType::TokAssign);
                    } else {
                        self.add_token(TokenType::TokColon);
                    }
                }
                ';' => self.add_token(TokenType::TokSemicolon),
                '?' => self.add_token(TokenType::TokQuestion),
                '>' => {
                    if self.match_char('=') {
                        self.add_token(TokenType::TokGe);
                    } else if self.match_char('>') {
                        self.add_token(TokenType::TokGtgt);
                    } else {
                        self.add_token(TokenType::TokGt);
                    }
                }
                '<' => {
                    if self.match_char('=') {
                        self.add_token(TokenType::TokLe);
                    } else if self.match_char('<') {
                        self.add_token(TokenType::TokLtlt);
                    } else {
                        self.add_token(TokenType::TokLt)
                    }
                }
                '=' => {
                    if self.match_char('=') {
                        self.add_token(TokenType::TokEq);
                    }
                }
                '~' => {
                    if self.match_char('=') {
                        self.add_token(TokenType::TokNe)
                    } else {
                        self.add_token(TokenType::TokNot);
                    }
                }
                '"' => self.handle_string(ch),
                '\'' => self.handle_string(ch),
                _ => {
                    if ch.is_ascii_digit() {
                        self.handle_number()
                    }
                    if ch.is_alphanumeric() {
                        self.handle_identifier()
                    }
                }
            }
        }
    }
}

File: zig/src/lexer.zig
const std = @import("std");
const tokens = @import("tokens.zig");
const ArrayList = std.ArrayList;
const ascii = std.ascii;

pub const Lexer = struct {
    tokens: ArrayList(tokens.Token),
    start: u32,
    curr: u32,
    line: u32,
    line_position: u32,
    source: []const u8,

    fn advance(self: *Lexer) u8 {
        if (self.curr >= self.source.len) {
            return ' ';
        }
        const ch = self.source[self.curr];
        self.curr += 1;
        self.line_position += 1;
        return ch;
    }

    fn add_token(self: *Lexer, token_type: tokens.TokenType) !void {
        try self.tokens.append(tokens.token_init(token_type, self.source[self.start..self.curr], self.line, self.line_position));
    }

    fn peek(self: *Lexer) u8 {
        if (self.curr >= self.source.len) {
            return ' ';
        }
        return self.source[self.curr];
    }

    fn lookahead(self: *Lexer) u8 {
        if (self.curr + 1 >= self.source.len) {
            return ' ';
        }
        return self.source[self.curr + 1];
    }

    fn match_char(self: *Lexer, expected: u8) bool {
        if ((self.curr >= self.source.len) or (self.source[self.curr] != expected)) {
            return false;
        }
        self.curr += 1;
        return true;
    }

    fn handle_number(self: *Lexer) !void {
        while (ascii.isDigit(self.peek())) {
            _ = self.advance();
        }
        if ((self.peek() == '.') and (ascii.isDigit(self.lookahead()))) {
            _ = self.advance();
            while (ascii.isDigit(self.peek())) {
                _ = self.advance();
            }
            try self.add_token(tokens.TokenType.TokFloat);
        } else {
            try self.add_token(tokens.TokenType.TokInteger);
        }
    }

    fn handle_string(self: *Lexer, quote: u8) !void {
        while ((self.peek() != quote) and (self.curr < self.source.len)) {
            _ = self.advance();
        }
        if (self.curr >= self.source.len) {
            std.debug.panic("Unfinished string at {}", .{self.line});
        }
        _ = self.advance();
        try self.add_token(tokens.TokenType.TokString);
    }

    fn handle_identifier(self: *Lexer) !void {
        while ((ascii.isAlphanumeric(self.peek())) or self.peek() == '_') {
            _ = self.advance();
        }
        try self.add_token(tokens.KeyWords.get(self.source[self.start..self.curr]) orelse tokens.TokenType.TokIdentifier);
    }

    pub fn tokenize(self: *Lexer) !void {
        while (self.curr < self.source.len) {
            self.start = self.curr;
            const ch = self.advance();
            switch (ch) {
                '\n' => {
                    self.line += 1;
                    self.line_position = 1;
                },
                '#' => {
                    while ((self.peek() != '\n') and (self.curr < self.source.len)) {
                        _ = self.advance();
                    }
                },
                '(' => try self.add_token(tokens.TokenType.TokLparen),
                ')' => try self.add_token(tokens.TokenType.TokRparen),
                '{' => try self.add_token(tokens.TokenType.TokLcurly),
                '}' => try self.add_token(tokens.TokenType.TokRcurly),
                '[' => try self.add_token(tokens.TokenType.TokLsquar),
                ']' => try self.add_token(tokens.TokenType.TokRsquar),
                ',' => try self.add_token(tokens.TokenType.TokComma),
                '.' => try self.add_token(tokens.TokenType.TokDot),
                '+' => try self.add_token(tokens.TokenType.TokPlus),
                '-' => {
                    if (self.match_char('-')) {
                        while (self.peek() != '\n' and self.curr < self.source.len) {
                            _ = self.advance();
                        }
                    } else {
                        try self.add_token(tokens.TokenType.TokMinus);
                    }
                },
                '*' => try self.add_token(tokens.TokenType.TokStar),
                '/' => try self.add_token(tokens.TokenType.TokSlash),

                '^' => try self.add_token(tokens.TokenType.TokCaret),
                '%' => try self.add_token(tokens.TokenType.TokMod),
                ':' => {
                    if (self.match_char('=')) {
                        try self.add_token(tokens.TokenType.TokAssign);
                    } else {
                        try self.add_token(tokens.TokenType.TokColon);
                    }
                },
                ';' => try self.add_token(tokens.TokenType.TokSemicolon),
                '?' => try self.add_token(tokens.TokenType.TokQuestion),
                '>' => {
                    if (self.match_char('=')) {
                        try self.add_token(tokens.TokenType.TokGe);
                    } else if (self.match_char('>')) {
                        try self.add_token(tokens.TokenType.TokGtgt);
                    } else {
                        try self.add_token(tokens.TokenType.TokGt);
                    }
                },
                '<' => {
                    if (self.match_char('=')) {
                        try self.add_token(tokens.TokenType.TokLe);
                    } else if (self.match_char('<')) {
                        try self.add_token(tokens.TokenType.TokLtlt);
                    } else {
                        try self.add_token(tokens.TokenType.TokLt);
                    }
                },
                '=' => {
                    if (self.match_char('=')) {
                        try self.add_token(tokens.TokenType.TokEq);
                    }
                },
                '~' => {
                    if (self.match_char('=')) {
                        try self.add_token(tokens.TokenType.TokNe);
                    } else {
                        try self.add_token(tokens.TokenType.TokNot);
                    }
                },
                '"' => try self.handle_string(ch),
                '\'' => try self.handle_string(ch),
                else => {
                    if (ascii.isDigit(ch)) {
                        try self.handle_number();
                    }
                    if (ascii.isAlphanumeric(ch)) {
                        try self.handle_identifier();
                    }
                },
            }
        }
    }
};

pub fn init_lexer(allocator: anytype, source: []const u8) !Lexer {
    try tokens.init_keywords(allocator);
    return .{ .tokens = ArrayList(tokens.Token).init(allocator), .start = 0, .curr = 0, .line = 1, .line_position = 1, .source = source };
}
